### PR TITLE
M/logic refactor

### PR DESCRIPTION
**Pull request by Todd-j-sutherland**
Description:

- created a new folder named data which houses different data sets for testing
- had to update the structure of the simulator to handle the csv strucutre
- at the end of the simulator we log the unsold (long and short) stocks
- in the zone_recovery we check to see if we make a profit with the sale if not it doesn't sell, this is a part that should be up for debate, I think it's good to not sell if there's no profit as eventually that stock might be in a position to handle a sale (eventually). but obviously this needs to be tested
- added a 100 day moving average this way we check not only 21 days for a trend but a 100 day moving in a positive

- [data.zip](https://github.com/user-attachments/files/15755792/data.zip) this is the files I've been testing, ideally have a forloop to test them all later on
 direction. I think this will help to prevent poor long term performance to sell at a loss
- removed some of the logging though feel free to add them back in, just attempting to remove the noise when testing

### CODE REVIEW

File: trading_simulation.py [hunk 1]
```diff
@@ -17,20 +17,22 @@ def load_stock_data_from_csv(file_path):
     chunks = pd.read_csv(file_path, chunksize=chunksize)
     df = pd.concat(chunks)
     print(f"Data loaded. Number of rows: {len(df)}")
-    # Combine 'Date' and 'Time' into a single 'Datetime' column
-    df['Datetime'] = pd.to_datetime(df['Date'] + ' ' + df['Time'])
+
+    # Remove 'GMT' and parse 'Local time' to datetime
+    df['Local time'] = df['Local time'].str.replace('GMT', '')
+    df['Datetime'] = pd.to_datetime(df['Local time'], format='%d.%m.%Y %H:%M:%S.%f %z')
```
Comment (saibottrenham) [inline trading_simulation.py:23]: format string would be good to have it as global variable

File: trading_simulation.py [hunk 2]
```diff
@@ -17,20 +17,22 @@ def load_stock_data_from_csv(file_path):
...
     print(f"First few rows of data:\n{df.head()}")
     df.set_index('Datetime', inplace=True)
     df.sort_index(inplace=True)
-    return df['Price']
+    return df['Close']
```
Comment (saibottrenham) [inline trading_simulation.py:27]: nice

File: zone_recovery_logic.py [hunk 3]
```diff
@@ -18,50 +17,73 @@ def __init__(self, rsi_period=22, entry_rsi_low=30, entry_rsi_high=72, profit_ta
...
+        
+        # Calculate indicators
         rsi = calculate_rsi(prices, self.rsi_period)
         ema21 = calculate_ema(prices, 21)
+        ema100 = calculate_ema(prices, 100)
         macd, signal = calculate_macd(prices)
+        macd_hist = macd - signal
+
+        # Ensure indicators are Series
+        current_rsi = rsi.iloc[-1] if isinstance(rsi, pd.Series) else rsi
```
Comment (saibottrenham) [inline zone_recovery_logic.py:29]: this might be worth a refactor. usually at this stage we should not check for type. It's better to rework the calculat_rsi functions etc to make sure they always return the expected type or None

File: zone_recovery_logic.py [hunk 4]
```diff
@@ -18,50 +17,73 @@ def __init__(self, rsi_period=22, entry_rsi_low=30, entry_rsi_high=72, profit_ta
...
-        stock_data['previous_rsi'] = rsi
+        stock_data['previous_rsi'] = current_rsi
 
         # Calculate total profit and individual losses
         total_profit = self.calculate_percentage_profit(stock_data['long'], stock_data['short'], current_price)
         long_loss = -self.calculate_percentage_profit(stock_data['long'], [], current_price)
         short_loss = -self.calculate_percentage_profit([], stock_data['short'], current_price)
 
-        # Check if profit target is reached to close all positions
+        logging.debug(f"{stock}: Current Price: {current_price}, RSI: {current_rsi}, Previous RSI: {previous_rsi}, EMA21: {current_ema21}, EMA100: {current_ema100}, MACD: {current_macd}, Signal: {current_signal}, Total Profit: {total_profit}%")
```
Comment (saibottrenham) [inline zone_recovery_logic.py:45]: good.

File: zone_recovery_logic.py [hunk 5]
```diff
@@ -18,50 +17,73 @@ def __init__(self, rsi_period=22, entry_rsi_low=30, entry_rsi_high=72, profit_ta
...
+
         trade_count = len(stock_data['long']) + len(stock_data['short'])
         if total_profit >= self.profit_target or trade_count >= self.max_trades:
-            logging.info(f"{stock}: Closing all positions due to reaching the profit target or max trades. Current profit: {total_profit}%.")
+            # logging.info(f"{stock}: Closing all positions due to reaching the profit target or max trades. Current profit: {total_profit}%.")
```
Comment (saibottrenham) [inline zone_recovery_logic.py:49]: should not comment out but when you run the simulation, just configure it that the loglevel is warning or error, That way we can keep the info logs in place but they won't get printed

**Comment by saibottrenham** [comment]
Looks good on the first pass. I will have a deeper look on Friday and give it a test run. I also gonna fix up the conflict while I'm at it.

File: zone_recovery_logic.py [hunk 6]
```diff
@@ -18,19 +17,35 @@ def __init__(self, rsi_period=22, entry_rsi_low=30, entry_rsi_high=72, profit_ta
 
     def calculate_indicators_and_check_profit(self, stock_data, stock, current_price):
```
Comment (saibottrenham) [inline zone_recovery_logic.py:18]: really good additions. The function itself might now becoming to big so it might be worth breaking that up into smaller pieces. 

For example -> 

def calculate_indicators_and_check_profit(self, stock_data, stock, current_price):
    prices = pd.Series(stock_data["prices"])

    # Calculate indicators
    indicators = self.calculate_indicators(prices)
    current_rsi = indicators['rsi']
    current_ema21 = indicators['ema21']
    current_ema100 = indicators['ema100']
    current_macd_hist = indicators['macd_hist']

    # Handle previous_macd_hist calculation
    previous_macd_hist = stock_data.get('previous_macd_hist')
    stock_data['previous_macd_hist'] = current_macd_hist

    previous_rsi = stock_data.get('previous_rsi')
    stock_data['previous_rsi'] = current_rsi

    # Calculate total profit and individual losses
    total_profit = self.calculate_percentage_profit(stock_data['long'], stock_data['short'], current_price)
    long_loss = -self.calculate_percentage_profit(stock_data['long'], [], current_price)
    short_loss = -self.calculate_percentage_profit([], stock_data['short'], current_price)

    logging.debug(f"{stock}: Current Price: {current_price}, RSI: {current_rsi}, Previous RSI: {previous_rsi}, EMA21: {current_ema21}, EMA100: {current_ema100}, Total Profit: {total_profit}%")

    trade_count = len(stock_data['long']) + len(stock_data['short'])
    if total_profit >= self.profit_target or trade_count >= self.max_trades:
        logging.info(f"{stock}: Closing all positions due to reaching the profit target or max trades. Current profit: {total_profit}%.")
        return "CLOSE_ALL", current_price, total_profit

    if long_loss > self.loss_threshold or short_loss > self.loss_threshold:
        return self.handle_loss(stock, long_loss, short_loss, current_price, total_profit)

    if previous_rsi is not None:
        rsi_signal = self.check_rsi_signals(stock, previous_rsi, current_rsi, current_price, total_profit)
        if rsi_signal:
            return rsi_signal

    return self.check_macd_ema_signals(stock_data, stock, current_price, current_ema21, current_ema100, current_macd_hist, previous_macd_hist, total_profit)

def calculate_indicators(self, prices):
    rsi = calculate_rsi(prices, self.rsi_period)
    ema21 = calculate_ema(prices, 21)
    ema100 = calculate_ema(prices, 100)
    macd, signal = calculate_macd(prices)
    macd_hist = macd - signal

    return {
        'rsi': rsi,
        'ema21': ema21,
        'ema100': ema100,
        'macd_hist': macd_hist
    }

def handle_loss(self, stock, long_loss, short_loss, current_price, total_profit):
    if long_loss > short_loss:
        logging.info(f"{stock}: Long positions showing greater loss at {long_loss}%. Hedging by opening a SHORT position at {current_price}.")
        return "SELL", current_price, total_profit
    else:
        logging.info(f"{stock}: Short positions showing greater loss at {short_loss}%. Hedging by opening a LONG position at {current_price}.")
        return "BUY", current_price, total_profit

def check_rsi_signals(self, stock, previous_rsi, current_rsi, current_price, total_profit):
    if current_rsi < self.entry_rsi_low and previous_rsi < current_rsi:
        logging.info(f"{stock}: RSI increasing from oversold ({previous_rsi} to {current_rsi}), triggered BUY at price {current_price}")
        return "BUY", current_price, total_profit
    elif current_rsi > self.entry_rsi_high and previous_rsi > current_rsi:
        logging.info(f"{stock}: RSI decreasing from overbought ({previous_rsi} to {current_rsi}), triggered SELL at price {current_price}")
        return "SELL", current_price, total_profit
    return None

def check_macd_ema_signals(self, stock_data, stock, current_price, current_ema21, current_ema100, current_macd_hist, previous_macd_hist, total_profit):
    if current_price > current_ema21 and current_price > current_ema100 and current_macd_hist > previous_macd_hist:
        logging.info(f"{stock}: Price above 21 EMA and 100 EMA and MACD histogram increasing. Triggered BUY at price {current_price}")
        return "BUY", current_price, total_profit
    elif current_price < current_ema21 and current_price < current_ema100 and current_macd_hist < previous_macd_hist:
        profitable_long_sell = any(pos['price'] < current_price for pos in stock_data['long'])
        if profitable_long_sell:
            logging.info(f"{stock}: Price below 21 EMA and 100 EMA and MACD histogram decreasing. Triggered profitable SELL at price {current_price}")
            return "SELL", current_price, total_profit

        profitable_short_cover = any(pos['price'] > current_price for pos in stock_data['short'])
        if profitable_short_cover:
            logging.info(f"{stock}: Price below 21 EMA and 100 EMA and MACD histogram decreasing. Triggered profitable COVER at price {current_price}")
            return "BUY", current_price, total_profit

    return None
